doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+  
+```python
+# No:
+def create_table(table_name):
+    # type: (str) -> Table
+    ...
+
+# Yes:
+def create_table(table_name: str) -> Table:
+    ...
+```
+
+- Do fully annotate function signatures - this includes type annotations for all parameters and the return type.
+- Do fully annotate function signatures in all samples - this ensures the type checks will check the sample code.
+- You should type annotate variables if the type in the code is different from expected, provides more value than what is already provided by Python itself, or if the type checker requires it.
+
+```python
+# No:
+table_name: str = "mytable"  # I can tell it's a string, not necessary
+create_table(table_name)
+
+# Yes:
+table_map: dict[str, Table] = {}  # clarifies what the dictionary expects
+table_map[table_name] = create_table(table_name)
+```
+
+- You do not need to type annotate `self` or `cls`.
+- Do return `None` from a constructor.
+  
+```python
+# No:
+class KeyCredential:
+
+    def __init__(self):
+        ...  # I do not get type checked
+
+# Yes:
+class KeyCredential:
+
+    def __init__(self) -> None:
+        ...
+```
+
+- Do use [mypy](https://mypy.readthedocs.io/en/stable/) and [pyright](https://github.com/microsoft/pyright) type checkers to statically type check your client library code.
+- Do use [black](https://pypi.org/project/black/) to format type annotations.
+- Do add type hints directly to the source code. If you think you need to use stub files, check with the architects. Note that with stub files, type checking will only work for _users_ of the stub, but won't type check the library code itself.
+- Do mark your client library package to distribute type hints according to [PEP 561](https://peps.python.org/pep-0561/).
+- Do use the latest typing features available. If not supported by older versions of Python, consider taking a dependency and importing from `typing-extensions`.
+
+```python
+# from typing import TypedDict Python >3.8
+from typing_extensions import TypedDict
+```
+
+### Importing types
+
+- Do use only publicly exposed client library types in type hints. 
+- Do import types from modules such as `typing`, `typing_extensions`, `collections`, and `collections.abc`(Note that indexing support for generic collection types from `collections.abc` is only supported on Python 3.9+).
+- Do not import regular type hints under a `typing.TYPE_CHECKING` block. You may use `TYPE_CHECKING` to fix a circular import or avoid importing a type only needed in type annotations that is otherwise costly to load at runtime.
+
+```python
+from typing import TYPE_CHECKING
+
+# No:
+if TYPE_CHECKING:
+    from typing import Union, TypeVar, Any
+
+
+# Yes:
+if TYPE_CHECKING:
+    from a import b  # avoiding a circular import
+    from expensive import c  # avoiding runtime costs
+```
+
+- Use `from __future__ import annotations` for forward declarations or when using built-in generic collection types, like `dict` or `list`.
```

Comment:
There are some caveats here. The Python community is torn about whether this will become the default at some point in the future or whether a different approach (PEP 649) will be used instead, at which point this form will be deprecated.

In general I would favor using this as a last resort -- if you have the occasional generic dict and you want to support Python 3.8 or before, it's probably better to use `typing.Dict[str, str]` instead of `dict[str, str]` + future import.